File: Day 5/day_5.py
f = open("./input.txt")
file = f.readlines()
f.close()


def get_row(input):
    max = 127
    min = 0
    for c in input:
        if(c == 'F'):
            max = int((min + max) / 2)
        elif(c == 'B'):
            min = int((min + max) / 2) + 1
        else:
            break
    return max


def get_column(input):
    max = 7
    min = 0
    for c in input:
        if(c == 'L'):
            max = int((min + max) / 2)
        elif(c == 'R'):
            min = int((min + max) / 2) + 1
        else:
            continue
    return max


ids = []

max = 0

for ticket in file:
    row = get_row(ticket)
    column = get_column(ticket)
    id = row * 8 + column
    if(id > max):
        max = id
    ids.append(id)

my_id = 0

for id1 in ids:
    has_next = False
    has_previous = False
    for id2 in ids:
        if id1 + 1 == id2:
            has_next = True
    if(not has_next and id1 != max):
        my_id = id1 + 1

print(f"Part 1: {max}\nPart 2: {my_id}")

File: Day 5/day_5.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>

using namespace std;

int GetRow(string ticket)
{
    int max = 127;
    int min = 0;

    for (int i = 0; i < ticket.size(); i++)
    {
        if (ticket[i] == 'F')
            max = (min + max) / 2;
        else if (ticket[i] == 'B')
            min = (min + max) / 2;
        else
            break;
    }
    return max;
}

int GetCol(string ticket)
{
    int max = 7;
    int min = 0;

    for (int i = 0; i < ticket.size(); i++)
    {
        if (ticket[i] == 'L')
            max = (min + max) / 2;
        else if (ticket[i] == 'R')
            min = (min + max) / 2;
        else
            continue;
    }
    return max;
}

int main()
{
    fstream f;
    f.open("input.txt");
    string l;
    vector<string> tickets;
    vector<int> ids;

    while (getline(f, l))
    {
        tickets.push_back(l);
        ids.push_back(GetRow(l) * 8 + GetCol(l));
    }
    f.close();

    int maxId = 0;
    int myId = 0;

    for (int i = 0; i < tickets.size(); i++)
    {
        bool hasNext = false;
        if (maxId < ids[i])
            maxId = ids[i];

        for (int j = 0; j < tickets.size(); j++)
        {
            if ((ids[i] + 1) == ids[j])
                hasNext = true;
        }

        if (!hasNext && ids[i] != maxId)
            myId = ids[i] + 1;
    }

    cout << "Part 1: " << maxId << "\nPart 2: " << myId << endl;

    return 0;
}
